src/interfaces/capnp/proxy-impl.h

```diff
@@ -0,0 +1,1733 @@
+    }
+}
+
+template <typename LocalTypes, typename... Args>
+void MaybeSetWant(LocalTypes, Priority<0>, Args&&...)
+{
+}
+
+template <typename Accessor, typename LocalType, typename ServerContext, typename Fn, typename... Args>
+void DefaultPassField(TypeList<LocalType>, ServerContext& server_context, Fn&& fn, Args&&... args)
+{
+    InvokeContext& invoke_context = server_context;
+    boost::optional<Decay<LocalType>> param;
+    const auto& params = server_context.call_context.getParams();
+    MaybeReadField(std::integral_constant<bool, Accessor::in>(), TypeList<LocalType>(), invoke_context,
+        Make<StructField, Accessor>(params), Emplace<decltype(param)>(param));
+    if (!param) param.emplace();
+    fn.invoke(server_context, std::forward<Args>(args)..., static_cast<LocalType&&>(*param));
+    auto&& results = server_context.call_context.getResults();
+    MaybeBuildField(std::integral_constant<bool, Accessor::out>(), TypeList<LocalType>(), BuildFieldPriority(),
+        invoke_context, *param, Make<StructField, Accessor>(results));
+}
+
+template <typename Output>
+void BuildField(TypeList<>,
+    Priority<1>,
+    InvokeContext& invoke_context,
+    Output&& output,
+    typename std::enable_if<std::is_same<decltype(output.get()), ThreadMap::Client>::value>::type* enable = nullptr)
+{
+    output.set(kj::heap<ProxyServer<ThreadMap>>(invoke_context.connection));
+}
+
+template <typename Input>
+void ReadFieldUpdate(TypeList<>,
+    InvokeContext& invoke_context,
+    Input&& input,
+    typename std::enable_if<std::is_same<decltype(input.get()), ThreadMap::Client>::value>::type* enable = nullptr)
+{
+    invoke_context.connection.m_thread_map = input.get();
+}
+
+template <typename Accessor, typename ServerContext, typename Fn, typename... Args>
+auto PassField(TypeList<>, ServerContext& server_context, const Fn& fn, Args&&... args) -> typename std::enable_if<
+    std::is_same<decltype(Accessor::get(server_context.call_context.getParams())), ThreadMap::Client>::value>::type
+{
+    const auto& params = server_context.call_context.getParams();
+    const auto& input = Make<StructField, Accessor>(params);
+    ReadFieldUpdate(TypeList<>(), server_context, input);
+    fn.invoke(server_context, std::forward<Args>(args)...);
+    auto&& results = server_context.call_context.getResults();
+    BuildField(TypeList<>(), BuildFieldPriority(), server_context, Make<StructField, Accessor>(results));
+}
+
+template <typename Derived, size_t N = 0>
+struct IterateFieldsHelper
+{
+    template <typename Arg1, typename Arg2, typename ParamList, typename NextFn, typename... NextFnArgs>
+    void handleChain(Arg1&& arg1, Arg2&& arg2, ParamList, NextFn&& next_fn, NextFnArgs&&... next_fn_args)
+    {
+        using S = Split<N, ParamList>;
+        handleChain(std::forward<Arg1>(arg1), std::forward<Arg2>(arg2), typename S::First());
+        next_fn.handleChain(std::forward<Arg1>(arg1), std::forward<Arg2>(arg2), typename S::Second(),
```

Comment:
Are `arg1` and `arg2` guaranteed to be valid here? Could have been moved from on the call on the line before?